### ISSUE TITLE
IdentityServer4 with Microsoft Accounts v2 endpoint

### ISSUE THREAD

**Issue by jimbo74** [IdentityServer/IdentityServer4.Samples#98]
Following the many examples for authenticating Identity Server 4 with google is easily done with the following code:

app.UseGoogleAuthentication(new GoogleOptions
            {
                AuthenticationScheme = "Google",
                DisplayName = "Google",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = "xxx",
                ClientSecret = "yyy"
            });

but trying to implement the microsoft v2 endpoint (so I can allow both work/school and personal accounts) using similar set up doesn't work.

            app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions
            {
                AuthenticationScheme = "Microsoft",
                DisplayName = "Microsoft",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = "vvvv"
                ClientSecret = "eeee"
            });

I have been unable to find a working example of this and I was hoping someone could please let me know if this is possible and, if so, how?

**Comment by leastprivilege** [comment]
That would be rather a question for Microsoft.

**Comment by jimbo74** [comment]
That might be so but they have examples of this set up that works in a simple Mvc app. It is only when I try to add the same set up to my identity server project it always fails for a personal account. I will ask for assistance at Microsoft instead then but I was just hoping someone here might have done this as I would think this is a common scenario.

**Comment by leastprivilege** [comment]
IdentityServer is just an MVC application - so it would be interesting to see the delta between your standalone app and identityserver.

..and what exactly does "fail" mean. Do you have logging enabled?

**Comment by jimbo74** [comment]
Sorry for my poor terminology, by fail I mean I enter my email at the MS sign in screen, I am then asked if it's a work or personal account. When I click on personal I am then prompted to enter a password. Following this I am presented with a screen that says:

Microsoft account is unavailable
Microsoft account is unavailable from this site, so you can't sign in or sign up. The site may be experiencing a problem.
You can sign in or sign up at other Microsoft sites and services, or try again later at this site.

The url of this page is:

https://login.live.com/ppsecure/post.srf?client_id=0d992256-6192-40c1-a280-bdde034c558e&scope=https%3A%2F%2Fgraph.microsoft.com%2Fuser.read&response_type=code&redirect_uri=https%3A%2F%2Flocalhost%3A44346%2Fsignin-microsoft&state=CfDJ8BKBaJzFexJKiWIQRg0MZq-bwEEQ4JIKHXfzktsrS-7OO8cHVAta3xH1mam7NOAtMlJ4JeZGYY8nxo6zFaSFGA0DaA-RXHMO3YOm_BZ6kj2X9qKrpi-WQNfEjcB2im_L726TTXJoYnZvpS8UKo-lD2cWpPzQZcfZGhKv08ZmsOLVCrtQ7SxIuwc-TEpbZDyYF2BauXidYAXowqbhY7jfNG66ywvY0o_QDQ7P5gmBa8Q_8DEKpoe2Xu9zAG_LJ1N5IZodsaYOOK0XvH_10xQHB3IOBn7ABZMbQKrtTQysr03w3NxpFOyvz_fDwEajD72F_T92mzX_5OKjEU75FFMwEeNVo1Q1WYB8CV6UPZ65Ymndr1tC9EGINStAmyqPihBdzn3OTUgWQSQ-Dsqa7gwgm1ykLuGBDl1Mgbe1kFjxm0AmaEMoJGBAWMJDhQCxZHVNjnDkcMxILXgKa7yhXkdSQhEDxYai1eY7_BVleUctdhXKlECsy3ezsBkyyU3CWh5OBOTrutfSHnRA82GNuXQDO9Jc0AD_4Mp5WQhOh--uIZc5a1bfRcoK_V5m1KM-Fs2NKnk5Z9-IzR-eX4vD6IisKJU5B-UccRrCI80hNiFuQms9-GL2BNOLMKlLVD1Qn-7KNSgX-lXhqrBNI0iX_hwHbUUp8qJSkD0PQN_YF3AJwl61YQEa7S1YxvLvHIaLAyfpGpEmBA66iQGyywycDlbI9FeREyeSl-EPEFkill21tEXyfeEwfoHLfDprLCzlLqI6SYaieSKl7248COk9baDbLYij3pTlrM_vbHq30MpDv_iSlARzOmZ1oVIKuX34fSEPaPKQFHUB-w1SYkrsRnINT2tMHF9oEYGfT8UP-Ab1X2YTqt12jTOHbK3cjDiiIAlPZ-B2nTBvOEpaOCCgBZ0xIEHGCIzUaSE9yrOya8Wz5yDtd7jsA5XKiPqdUGXwWBDug9MdLoU2KMeM_9Ez4SmKIOFbgt2iN5FLoJ4bbVWa4sL9g4IiS4pDO6FY1oVku9B4F92JL_amJcuLbPZFqAI0cmbUG5txnxDYJMuey7aI-JHiq7DaSUvF2GN_2ws_nxyiMxqL1rSeVnI1STfOOOoTz1hJKRyWt4FqaRuiST6urdMm10aYRqhQhhFcxKZk6BakqnMwJhR9Rx-UNS8RrY5I5_DvAOjeOo57dFT7Oa7TQBGoBcDSrHQFE9MSsr7h7nvMa4fOV9RaURBTZmC9alskqN8orQRE9D_5A9VIk5DNQ61qGfj1TZq-bz6ccp41mkQyzaz4O17seNRPjungZ9hpEiOA9qAJPYMo6y56uo0BhUYGI3KO_7d62aQBVm0jchTGyVbnGipR2anImzyp4dXyhkKPCLR0LUu0HWLQinIEM7CA9pz0gEfADDVdGQUvXjErn4QurLrCi8zJaSLWqplU60S-nLjhc2c6X5iofcOMxkeU&login_hint=my_email_removed&ui_locales=en-US&display=touch&issuer=mso&msproxy=1&contextid=344B5C959FCC0F21&bk=1481712992&uaid=51c7c98976f242199efa6e63895e3717&pid=15216


My log only gets as far as this:

Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action IdentityService.Controllers.AccountController.Login (IdentityService) in 1138.9268ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 1306.169ms 200 text/html; charset=utf-8
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request starting HTTP/1.1 POST http://localhost:44346/Account/ExternalLogin?returnurl=%2Fconnect%2Fauthorize%2Flogin%3Fclient_id%3Dmatrixui%26redirect_uri%3Dhttps%253A%252F%252Flocalhost%253A44379%252Fsignin-oidc%26response_type%3Dcode%2520id_token%26scope%3Dopenid%2520profile%2520grantapi%2520hubapi%2520offline_access%2520userdetails%26response_mode%3Dform_post%26nonce%3D636173097736189336.N2RhMTIzMTAtODU0NC00MDBjLWFmYWYtYmZlNTE0ODdkNTY3MzNmNmQ0ZDAtYzgzOC00ZWI0LTg3ODMtM2Q1Zjg5ODdhYmYz%26state%3DCfDJ8BKBaJzFexJKiWIQRg0MZq87UFTzEH9YDpOPiBbeVkhdXyuN-QjWtJfLLc3lN1Ugp7ZGYgWLo0Y9OyyGOyDaFfJEeyvmcUBhNW5D6UjcjPT3ShJvqqzY_-z6mAho2gux0bsPN-A5xB2AOC8NLaFfLOgx6-66-XCdQZHjZDweDoU2hvc5lerm6eAZKK3_l077WTZC2Z4htowYBMUj1Z22sXY83PQNB1m4Hd9IKl_gYlsv-5ZE_i7WX2RmkeX14Zw4cO-keIH9kTw9BefFXpexD-ueKWPg1xZgWWX3Kbp1FTrVLJlaysaNy1d08J-POrIVcKrCKTv4IO9dDG--6CaDBTgnf_iHh9jhs2Ouw4FoDXY0A3MiqwJmok__BjWkkmYfz8oGu7yxAnHrnbOyKNZvNMY application/x-www-form-urlencoded 201
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executing action method IdentityService.Controllers.AccountController.ExternalLogin (IdentityService) with arguments (Microsoft, /connect/authorize/login?client_id=matrixui&redirect_uri=https%3A%2F%2Flocalhost%3A44379%2Fsignin-oidc&response_type=code%20id_token&scope=openid%20profile%20grantapi%20hubapi%20offline_access%20userdetails&response_mode=form_post&nonce=636173097736189336.N2RhMTIzMTAtODU0NC00MDBjLWFmYWYtYmZlNTE0ODdkNTY3MzNmNmQ0ZDAtYzgzOC00ZWI0LTg3ODMtM2Q1Zjg5ODdhYmYz&state=CfDJ8BKBaJzFexJKiWIQRg0MZq87UFTzEH9YDpOPiBbeVkhdXyuN-QjWtJfLLc3lN1Ugp7ZGYgWLo0Y9OyyGOyDaFfJEeyvmcUBhNW5D6UjcjPT3ShJvqqzY_-z6mAho2gux0bsPN-A5xB2AOC8NLaFfLOgx6-66-XCdQZHjZDweDoU2hvc5lerm6eAZKK3_l077WTZC2Z4htowYBMUj1Z22sXY83PQNB1m4Hd9IKl_gYlsv-5ZE_i7WX2RmkeX14Zw4cO-keIH9kTw9BefFXpexD-ueKWPg1xZgWWX3Kbp1FTrVLJlaysaNy1d08J-POrIVcKrCKTv4IO9dDG--6CaDBTgnf_iHh9jhs2Ouw4FoDXY0A3MiqwJmok__BjWkkmYfz8oGu7yxAnHrnbOyKNZvNMY) - ModelState is Valid
Microsoft.AspNetCore.Mvc.ChallengeResult:Information: Executing ChallengeResult with authentication schemes (Microsoft).
Microsoft.AspNetCore.Authentication.MicrosoftAccount.MicrosoftAccountMiddleware:Information: AuthenticationScheme: Microsoft was challenged.
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker:Information: Executed action IdentityService.Controllers.AccountController.ExternalLogin (IdentityService) in 12.2814ms
Microsoft.AspNetCore.Hosting.Internal.WebHost:Information: Request finished in 21.8603ms 302 

The only difference I can see between the working example and my identityserver implementation is the working example only has:

 app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions()
            {
                ClientId = "someid",
                ClientSecret = "somesecret"
            });

where my identityserver has:

 app.UseCookieAuthentication(new CookieAuthenticationOptions
            {
                AuthenticationScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                AutomaticAuthenticate = false,
                AutomaticChallenge = false
            });
            app.UseFacebookAuthentication(new FacebookOptions()
            {
                AuthenticationScheme = "Facebook",
                DisplayName = "Facebook",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                AppId = Configuration["Authentication:Facebook:AppId"],
                AppSecret = Configuration["Authentication:Facebook:AppSecret"],
            });
            app.UseGoogleAuthentication(new GoogleOptions
            {
                AuthenticationScheme = "Google",
                DisplayName = "Google",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = Configuration["Authentication:Google:AppId"],
                ClientSecret = Configuration["Authentication:Google:AppSecret"]
            });
            app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions
            {
                AuthenticationScheme = "Microsoft",
                DisplayName = "Microsoft",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = Configuration["Authentication:MS:AppId"],
                ClientSecret = Configuration["Authentication:MS:AppSecret"]
            });

thanks in advance for any help you can offer

**Comment by jimbo74** [comment]
So this has completely thrown me as it now works if I change my IdentityServer startup (remove Facebook & Google) so I just have:

app.UseCookieAuthentication(new CookieAuthenticationOptions
{
AuthenticationScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
AutomaticAuthenticate = false,
AutomaticChallenge = false
});
app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions
{
AuthenticationScheme = "Microsoft",
DisplayName = "Microsoft",
SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
ClientId = Configuration["Authentication:MS:AppId"],
ClientSecret = Configuration["Authentication:MS:AppSecret"]
});

**Comment by tomresing** [comment]
Did you ever find a solution to this issue? I see the Microsoft account is unavailable message in a somewhat similar scenario.

**Comment by jimbo74** [comment]
No I'm afraid I couldn't get it to reliably work for me at the time so I went back to using OpenIdConnectAuthentication for microsoft accounts along the lines of what is shown below. Note that as this is using the v2 endpoint I still get errors when logging in with a personal account.
Sorry for not being able to offer anymore help.

app.UseOpenIdConnectAuthentication(new OpenIdConnectOptions
            {
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                AuthenticationScheme = "AAD",
                DisplayName = "Microsoft",
                Authority = "https://login.microsoftonline.com/Common/v2.0",
                ClientId = myclientid,
                ClientSecret = mysecret,
                ResponseType = "id_token",
                PostLogoutRedirectUri = mylogouturi,
                },
                Scope =  {
                            "openid", "profile"
                        }
            });

**Comment by tomresing** [comment]
Thanks for the response and detail of your work-around.

**Comment by jimbo74** [comment]
Hi tom just to let you know this now works with the Microsoft middleware. I don't know when they added it but I looked into this again last week and noticed that in their app set up they now allow for a redirect URL & when you populate it correctly you end up with a complete login/logout on the v2 endpoint. Take another look at it & if you need anymore help I'll happily assist if I xan

**Comment by kanvishok** [comment]
Hello everyone, 

Anyone solved this issue without any workaround?

Thanks

**Comment by jimbo74** [comment]
Yes I have it working, I can get you a code sample tomorrow

**Comment by kanvishok** [comment]
app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions()
            {
                AuthenticationScheme = "Office 365",
                DisplayName = "Office 365",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = "xxxx",
                ClientSecret = "xx"
            });

**Comment by kanvishok** [comment]
Thant would be great. Thanks Jimbo

**Comment by jimbo74** [comment]
ok I have this working but it does seem temperamental as I was happy with the way it was working last week, come back to test it today and it wasn't working again but now it is. When you get an error informing you that Microsoft is having trouble signing you in right now, they aren't joking. As Dominick said at the top of this issue, that's a Microsoft problem though not IS.
My code, which is not that much different to yours, and is currently working is as below:

 app.UseMicrosoftAccountAuthentication(new MicrosoftAccountOptions
            {
                AuthenticationScheme = "MS",
                DisplayName = "Microsoft",
                SignInScheme = IdentityServerConstants.ExternalCookieAuthenticationScheme,
                ClientId = Configuration["Authentication:Microsoft:AppId"],
                ClientSecret = Configuration["Authentication:Microsoft:AppSecret"]
            });

AppId & AppSecret point to an application set up in (https://apps.dev.microsoft.com) and there I have the following settings:
Allow Implicit flow = true
Redirect URLs = https://identityserverimplementation.net/signin-microsoft
Logout URL = https://identityserverimplementation.net/
Microsoft Graph Permissions = User.Read

and that's it working.. I hope this helps!

**Comment by andriysavin** [comment]
Let me share my experience with this. My deployment has several services, of them one is a UI web app and one is a web app hosting Identity Server. The apps are hosted in Azure. The IS is configured with external logins, one of them is Microsoft.

My flow: When an unauthenticated user comes to the UI app, he gets redirected to the IS (storing callback URI in the query string). When the user clicks "Microsoft account" button on the IS login page, he gets redirected again, now to the MS authorization server (storing one more callback URI in the query string). Then, when the user tries to sign in with his login/pass the error being discussed happens.

Investigating this problem I noticed that the same account was able to login when initially navigating directly to IS login page (skipping UI app). Since the only difference in this case (from MS auth service perspective) is some additional UI-web-app-related state in the query string (which MS service ignores), I supposed that the error is caused with the too long URI. Actually, before that I had solved problem with too long URIs on the IS hosting edge (Azure IIS URI filtering module was rejecting long URIs).
Indeed, my UI app URI was looking like https://xxxxxxxxxxxxxxxxxxxui.azurewebsites.net, and the IS URI  looking like https://xxxxxxxxxxxxxxxxxxxidentityserver.azurewebsites.net . 

I started with renaming IS service domain name to something like https://xxxxidsrv.azurewebsites.net, and that was enough for the error to disappear. But to be sure I would rename UI service as well.  Also, azurewebsites.net domain is not the shortest in the world, so switching to company custom domain (which is shorter in my case) will help too.
